### PR TITLE
Added auth data source

### LINKED ISSUE #184: Implement AuthDataSource
Login/Logout
Store auth data
Provide updates about auth state

### PR DESCRIPTION
**Pull request by antonshilov**
**Description**:
Added auth datasource

**Related issue(s)**:
Closes #184

### CODE REVIEW

File: android/app-example/src/main/java/com/badoo/ribs/example/auth/AuthDataSource.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        )
+            .doOnSuccess(::onAuthSuccess)
+
+    override fun loginAnonymous() {
+        val state = AuthState.Anonymous
+        updateState(state)
+    }
+
+    override fun logout() {
+        val state = AuthState.Anonymous
```
Comment (zsoltk) [inline android/app-example/src/main/java/com/badoo/ribs/example/auth/AuthDataSource.kt:43]: Shouldn't this be `AuthState.Unauthenticated`?

File: android/app-example/src/main/java/com/badoo/ribs/example/auth/AuthDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        updateState(state)
+    }
+
+    override fun getState(): AuthState =
+        if (stateRelay.hasValue()) {
+            stateRelay.value!!
+        } else throw IllegalStateException("Cannot retrieve auth state")
```
Comment (zsoltk) [inline android/app-example/src/main/java/com/badoo/ribs/example/auth/AuthDataSource.kt:50]: Fall back on Unauthenticated and log error? As I understand this branch shouldn't ever happen with the current implementation of the storage, but it's maybe better if there's no reliance on that.

File: android/app-example/src/main/java/com/badoo/ribs/example/auth/AuthDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    private fun onAuthSuccess(token: AccessToken) {
```
Comment (zsoltk) [inline android/app-example/src/main/java/com/badoo/ribs/example/auth/AuthDataSource.kt:52]: I'd suggest moving this closer to where it's used.

File: android/app-example/src/main/java/com/badoo/ribs/example/auth/AuthDataSource.kt (hunk 1 above)
Comment (antonshilov) [reply android/app-example/src/main/java/com/badoo/ribs/example/auth/AuthDataSource.kt:43]: It depends on the behavior we want get. Do we want a user to be redirected to the Welcome screen again or to the feed but as an anonymous user?

Comment (zsoltk) [reply android/app-example/src/main/java/com/badoo/ribs/example/auth/AuthDataSource.kt:43]: As a user, I'd personally expect to be thrown to welcome screen. We can always change this as we go.

**Pull request merged**